src/tactic/binder_matching.lean

```diff
@@ -0,0 +1,282 @@
+@[inline] private meta def mk_binders_aux {α}
+  (match_binder : ℕ → expr → tactic (option (name × binder_info × expr × expr)))
+  (binder_replacement : name → binder_info → expr → tactic expr)
+  (result : ℕ → name → binder_info → expr → expr → tactic (option α)) :
+  ℕ → expr → tactic (list α × expr) :=
+λ depth e, do
+  some (name, bi, type, body) ← match_binder depth e | pure ([], e),
+  replacement ← binder_replacement name bi type,
+  oa ← result depth name bi type body,
+  (as, rest) ← mk_binders_aux (depth + 1) (body.instantiate_var replacement),
+  let as' := oa.elim as (λ a, a :: as),
+  pure (as', rest)
+
+/--
+`mk_binders` implements the common functionality of functions like
+`get_pi_binders` and `mk_meta_pis`. It proceeds as follows:
+
+1. Match a pi or lambda binder using `match_binder`. `match_binder` should
+   return the name, binder_info, type and body of a leading binder.
+   Stop if this returns `none`.
+2. Use `binder_replacement` to constructs a replacement for this binder (in
+   our applications either a local constant or a metavariable) and instantiate
+   the bound variable with this replacement in the binder body.
+3. Use `result` to construct a result.
+4. Recurse into the binder body.
+
+Returns the list of results and the rest of the expression (with previously
+bound variables instantiated with their replacements).
+
+`binder_replacement` and `result` receive the binder information returned by
+`match_binder`. `match_binder` and `result` also receive the current recursion
+depth.
+
+If `match_binder`, `binder_replacement` or `result` fail at any point, the whole
+tactic fails.
+-/
+@[inline] meta def mk_binders {α}
+  (match_binder : ℕ → expr → tactic (option (name × binder_info × expr × expr)))
+  (binder_replacement : name → binder_info → expr → tactic expr)
+  (result : ℕ → name → binder_info → expr → expr → tactic (option α)) :
+  expr → tactic (list α × expr) :=
+mk_binders_aux match_binder binder_replacement result 0
+
+/--
+A special case of `mk_binders` which returns the binder replacements returned by
+`binder_replacement`.
+-/
+@[inline] meta def mk_binders'
+  (match_binder : ℕ → expr → tactic (option (name × binder_info × expr × expr)))
+  (binder_replacement : name → binder_info → expr → tactic expr) :
+  expr → tactic (list expr × expr) :=
+mk_binders match_binder binder_replacement
+  (λ _ name bi type _, some <$> binder_replacement name bi type)
+
+/--
+Auxiliary function which is used by the `mk_{local,meta}_{pis,lambdas}n` family
+of functions. It implements the "match exactly `max_depth` binders" logic.
+-/
+@[inline] private meta def match_with_depth {α}
+  (match_binder : expr → tactic (option α))
+  (max_depth : ℕ) (current_depth : ℕ) (e : expr) : tactic (option α) :=
+  if current_depth ≥ max_depth then none else do
+    (some x) ← match_binder e | failed,
+    pure (some x)
+
+/--
+Like `tactic.mk_local_pis` but the Π binders of the input expression are
```

Comment:
Maybe it's better to keep these doc strings standalone, without reference to previous doc strings. Much easier to understand the function without having to look up other functions. You can and should reference `tactic.mk_local_pis` at the end, e.g. `See also ...`.

Unfortunately this request applies in many of the doc strings here. Sorry to ask for the extra busy work.

Reply:
Done. I hope I haven't made too many copy-and-paste errors.